--- insert_with_mutation.ml.orig	2019-01-31 15:14:50.160015885 +0000
+++ insert_with_mutation.ml	2019-01-31 15:15:43.273333629 +0000
@@ -363,7 +363,15 @@ let rec make_initial_find_state k r = F_
 
 end;; (*struct Pre_monad*)
 
-module Insert_with_mutation : sig
+module type MONAD = sig
+  type t
+  type ('a,'t) mm = ('a,t) Tjr_monad.Types.m
+  val bind : ('a -> ('b, 'c) mm) -> ('a, 'c) mm -> ('b, 'c) mm
+  val fmap : ('a -> 'b) -> ('a, 'c) mm -> ('b, 'c) mm
+  val return : 'a -> ('a, 'b) mm
+end
+
+module Insert_with_mutation(Monad:MONAD) : sig
   type ('a, 'b, 'c) blk_ops =
     Make_blk_ops of
       ('a -> ('b, 'c) Monad.mm) * ('b -> ('a, 'c) Monad.mm) *
