--- insert_with_mutation.ml.orig	2019-02-03 12:31:36.798069584 +0000
+++ insert_with_mutation.ml	2019-02-03 12:32:06.294727924 +0000
@@ -373,7 +373,15 @@ let rec make_initial_find_state k r = F_
 
 end;; (*struct Pre_monad*)
 
-module Insert_with_mutation : sig
+module type MONAD = sig
+  type t
+  type ('a,'t) mm = ('a,t) Tjr_monad.Types.m
+  val bind : ('a -> ('b, 'c) mm) -> ('a, 'c) mm -> ('b, 'c) mm
+  val fmap : ('a -> 'b) -> ('a, 'c) mm -> ('b, 'c) mm
+  val return : 'a -> ('a, 'b) mm
+end
+
+module Insert_with_mutation(Monad:MONAD) : sig
   type ('a, 'b, 'c) store_ops =
     Make_store_ops of
       ('a -> ('b, 'c) Monad.mm) * ('b -> ('a, 'c) Monad.mm) *
